### DISCUSSION TITLE
I ran `prisma db push`, and now when I run `prisma migrate dev` it prompts me to reset the database. How do I migrate without resetting and losing the data?

### DISCUSSION THREAD

**Discussion by lumenwrites** [prisma/prisma#16141] [Q&A] [answered]
Hi! When I run `prisma migrate dev` I see the error:
```
Drift detected: Your database schema is not in sync with your migration history.

The following is a summary of the differences between the expected database schema given your migrations files, and the actual schema of the database.

It should be understood as the set of changes to get from the expected schema to the actual schema.

[+] Added tables
  - Comment
  - Post
  - _ProfileUpvotedComments
  - _ProfileUpvotedPosts

[*] Changed the `Comment` table
  [+] Added foreign key on columns (authorId)
  [+] Added foreign key on columns (parentId)
  [+] Added foreign key on columns (postId)

[*] Changed the `Post` table
  [+] Added unique index on columns (slug)
  [+] Added foreign key on columns (authorId)
  [+] Added foreign key on columns (lessonId)

[*] Changed the `Profile` table
  [+] Added column `username`
  [+] Added unique index on columns (username)

[*] Changed the `_ProfileUpvotedComments` table
  [+] Added unique index on columns (A, B)
  [+] Added index on columns (B)
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)

[*] Changed the `_ProfileUpvotedPosts` table
  [+] Added unique index on columns (A, B)
  [+] Added index on columns (B)
  [+] Added foreign key on columns (A)
  [+] Added foreign key on columns (B)

✔ We need to reset the PostgreSQL database "postgres" at "db.qqdrlwubvxmhbwsddvlx.supabase.co:5432".
Do you want to continue? All data will be lost. … no
```
I'm not sure why it happened, but I think it might've happened because I ran `prisma db push` as I was developing a new featrue, without understanding that this will result in me having to reset the database.

How can I migrate without resetting the database? I need to keep the database and the data.
And how can I avoid running into this question in the future? I need to make sure that I won't have to reset the database and lose the data after the website has been launched.

**Comment by pantharshit00** [comment, marked as answer]
Hello @lumenwrites 

`npx prisma db push` is meant for development use only. The development db is considered "resettable" and "seedable" as its used to develop your migrations and to test/tweak them if needed. Looks like you have used it in your production environment. In production, you should use `npx prisma migrate deploy` and it would ensure your db won't get reset in any case. 


Now in order to resolve your current situation we need to tweak the migrations history using some primitives which migrate provides like the `diff` command(https://www.prisma.io/docs/reference/api-reference/command-reference#migrate-diff)

1. You will need to generate a new migration which will be the difference between migrations history stored in your disk in the migrations folder and your production database. The `--to-schema-datasource` flag here point to the database stored in your `.env` file but you can also change this flag to `--to-url` if you prefer to give the database url in the command itself. 
```
mkdir -p prisma/migrations/add-comments-and-post/
npx prisma migrate diff --from-migrations prisma/migrations --to-schema-datasource prisma/schema.prisma --shadow-database-url <url-to-any-empty-pg-database>  --script >  prisma/migrations/add-comments-and-post/migration.sql
```

The `--shadow-database-url` parameter is required and you will need to point it to any empty postgres database either a local one or hosted one. Prisma will use that to reply migration history and calculate the diff. 

This will create a new migration in `prisma/migrations/add-comments-and-post/migration.sql` which will have the SQL for the model you have added via `db push`. 


2. Now you have the migration ready, we will need to mark it as applied. This is because those tables already exists in the database and we can't run the actual `CREATE` sql statements to the database. We will just mark the migration as applied as it was already applied when you ran `npx prisma db push`. 

```
npx prisma migrate resolve --applied add-comments-and-post
```

3. Now, you should be ready to go. Run `npx prisma migrate dev` once now and it should say everything is in sync.

**Reply by Veercodeprog** [reply]
@pantharshit00 can you describe it ,i am currently using prisma for nodejs app , every time i make changes to schema and migrate ,i lost the database values.how can i avoid it. i am avoiding using push and migrate at any cost,but then changes in db doesnt reflects in prisma queries

**Comment by Veercodeprog** [comment]
found alternate simple answer,



Go into your schema and make the change:

NormalizedName  String            @unique @db.VarChar(64)
NormalizedName  String?            @unique @db.VarChar(64)
Then create a draft migration:

$ npx prisma migrate dev --name migration-name --create-only
Then edit the migration in SQL (this allow null values ie optional)

ALTER TABLE myTable ALTER COLUMN myColumn {DataType} NULL;
OR Postgres

ALTER TABLE myTable ALTER COLUMN myColumn DROP NOT NULL;
etc.

Then apply the modified SQL:

$ npx prisma migrate dev

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!

**Comment by itspers** [comment]
This is extremelly strange. After two years developing with Prisma i just discovered this 'db push' and thought 'Ah, so thats how you supposed to work all this time'
So today started another big task, made this push and expected to make one nice migration in the end. And boom - needs to be reset 😮 
Like what a sense of it then? I would just make migration to every 10 fields i add like before and dont have this another headache.

`The development db is considered "resettable" and "seedable" as its used to develop your migrations and to test/tweak them if needed`

What kind of development it is, some hello world app...
